File: experiments/java-sdk/src/main/java/com/datadog/llmobs/experiments/Experiment.java
package com.datadog.llmobs.experiments;

import com.datadog.api.client.v2.api.LlmObservabilityApi;
import com.datadog.api.client.v2.model.LLMObsExperimentDataAttributesRequest;
import com.datadog.api.client.v2.model.LLMObsExperimentDataRequest;
import com.datadog.api.client.v2.model.LLMObsExperimentRequest;
import com.datadog.api.client.v2.model.LLMObsExperimentResponse;
import com.datadog.api.client.v2.model.LLMObsExperimentType;
import com.datadog.llmobs.experiments.internal.DirectPost;
import com.datadog.llmobs.experiments.internal.MetricBuilder;
import com.datadog.llmobs.experiments.internal.SpanBuilder;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

/**
 * Runs a user's {@link Task} against every record of a {@link Dataset}, executes
 * {@link Evaluator}s on each output, and ships spans + eval metrics to Datadog.
 *
 * <p>Build via {@link #builder(ExperimentsClient)}. Execute with {@link #run()}.
 *
 * <p>v0.1: sequential execution only. Parallelism is a v0.2 add.
 */
public final class Experiment<I, O> {

    private final ExperimentsClient client;
    private final String name;
    private final String description;
    private final Dataset dataset;
    private final Task<I, O> task;
    private final Map<String, Evaluator<?>> evaluators;
    private final Map<String, Object> config;
    private final Map<String, String> tags;

    private String experimentId;

    private Experiment(Builder<I, O> b) {
        this.client = b.client;
        this.name = b.name;
        this.description = b.description == null ? "" : b.description;
        this.dataset = b.dataset;
        this.task = b.task;
        this.evaluators = new LinkedHashMap<>(b.evaluators);
        this.config = b.config == null ? Map.of() : Map.copyOf(b.config);
        this.tags = b.tags == null ? Map.of() : Map.copyOf(b.tags);
    }

    public static <I, O> Builder<I, O> builder(ExperimentsClient client) {
        return new Builder<>(client);
    }

    public String name() { return name; }
    public String experimentId() { return experimentId; }

    public String url() {
        if (experimentId == null) {
            return null;
        }
        return "https://app." + client.site() + "/llm/experiments/" + experimentId;
    }

    @SuppressWarnings("unchecked")
    public ExperimentResult run() {
        // Shutdown hook updates the experiment status to "interrupted" if the JVM is
        // killed (Ctrl-C, SIGTERM, System.exit) while run() is in flight. No-op until
        // experimentId is populated. Removed in the finally block on clean exit.
        Thread interruptHook = new Thread(() -> {
            if (this.experimentId != null) {
                try {
                    updateStatus("interrupted", "Process terminated before run completed");
                } catch (Exception ignored) {
                    // best effort during JVM shutdown
                }
            }
        }, "experiment-interrupt-hook");
        Runtime.getRuntime().addShutdownHook(interruptHook);

        String finalStatus = "completed";
        String finalError = null;

        try {
            LlmObservabilityApi api = client.api();

            String projectId = client.ensureProjectId();
            dataset.ensureCreatedAndPushed(projectId);

            LLMObsExperimentRequest expReq = new LLMObsExperimentRequest(
                new LLMObsExperimentDataRequest()
                    .type(LLMObsExperimentType.EXPERIMENTS)
                    .attributes(
                        new LLMObsExperimentDataAttributesRequest(name, projectId)
                            .datasetId(dataset.id())
                            .description(description)
                            .config(config.isEmpty() ? null : new LinkedHashMap<>(config))
                            .ensureUnique(true)
                    )
            );

            LLMObsExperimentResponse expResp;
            try {
                expResp = api.createLLMObsExperiment(expReq);
            } catch (Exception e) {
                throw new RuntimeException("Failed to create experiment '" + name + "': " + e.getMessage(), e);
            }
            this.experimentId = expResp.getData().getId();

            List<ExperimentResult.Row> rows = new ArrayList<>();
            List<Map<String, Object>> spans = new ArrayList<>();
            List<Map<String, Object>> metrics = new ArrayList<>();

            List<DatasetRecord> records = dataset.records();
            List<String> recordIds = dataset.recordIds();
            int idx = 0;
            for (DatasetRecord record : records) {
                String recordId = idx < recordIds.size() ? recordIds.get(idx) : "";
                String spanId = UUID.randomUUID().toString().replace("-", "");
                String traceId = UUID.randomUUID().toString().replace("-", "");
                long startNs = System.currentTimeMillis() * 1_000_000L;
                long startMonoNs = System.nanoTime();

                Object output = null;
                String errType = null;
                String errMsg = null;
                try {
                    output = task.execute((I) record.input(), config);
                } catch (Exception e) {
                    errType = e.getClass().getName();
                    errMsg = e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage();
                }

                long durationNs = System.nanoTime() - startMonoNs;
                long endTimestampMs = (startNs + durationNs) / 1_000_000L;

                Map<String, Object> evalValues = new LinkedHashMap<>();
                Map<String, String> evalErrors = new LinkedHashMap<>();
                for (Map.Entry<String, Evaluator<?>> e : evaluators.entrySet()) {
                    try {
                        Object v = e.getValue().evaluate(record.input(), output, record.expectedOutput());
                        evalValues.put(e.getKey(), v);
                    } catch (Exception ex) {
                        String msg = ex.getMessage() == null ? ex.getClass().getSimpleName() : ex.getMessage();
                        evalErrors.put(e.getKey(), msg);
                        evalValues.put(e.getKey(), null);
                    }
                }

                ExperimentResult.Row row = new ExperimentResult.Row(
                    idx, spanId, traceId, startNs, durationNs,
                    record.input(), output, record.expectedOutput(),
                    errType, errMsg, evalValues, evalErrors
                );
                rows.add(row);

                spans.add(SpanBuilder.toSpan(row, record.metadata(), projectId, dataset.id(), recordId, experimentId, name, tags));
                for (Map.Entry<String, Object> eval : evalValues.entrySet()) {
                    metrics.add(MetricBuilder.toMetric(
                        eval.getKey(), eval.getValue(), evalErrors.get(eval.getKey()),
                        spanId, endTimestampMs, experimentId, tags
                    ));
                }
                idx++;
            }

            EventsPoster.post(client, experimentId, spans, metrics);

            return new ExperimentResult(experimentId, rows, url());
        } catch (RuntimeException | Error e) {
            finalStatus = "failed";
            String msg = e.getMessage();
            finalError = msg != null ? msg : e.getClass().getSimpleName();
            throw e;
        } finally {
            try {
                Runtime.getRuntime().removeShutdownHook(interruptHook);
            } catch (IllegalStateException ignored) {
                // JVM is already shutting down; the hook will run on its own
            }
            if (this.experimentId != null) {
                try {
                    updateStatus(finalStatus, finalError);
                } catch (Exception ignored) {
                    // never mask the original exception with a status-update failure
                }
            }
        }
    }

    /**
     * PATCH the experiment with a new status (and optional error message). Hand-rolled HTTP:
     * the generated {@code LLMObsExperimentUpdateDataAttributesRequest} only exposes {@code name}
     * and {@code description} — the spec doesn't surface {@code status} / {@code error} despite
     * the live API accepting them. Once the spec catches up, this can be replaced with
     * {@code api.updateLLMObsExperiment(...)}.
     */
    private void updateStatus(String status, String errorMessage) {
        Map<String, Object> attributes = new HashMap<>();
        attributes.put("status", status);
        if (errorMessage != null) {
            attributes.put("error", errorMessage);
        }

        Map<String, Object> data = new HashMap<>();
        data.put("type", "experiments");
        data.put("id", experimentId);
        data.put("attributes", attributes);

        Map<String, Object> body = new HashMap<>();
        body.put("data", data);

        String path = "/api/v2/llm-obs/v1/experiments/" + experimentId;
        DirectPost.patch(client.site(), path, client.apiKey(), client.applicationKey(), body);
    }

public static final class Builder<I, O> {
        private final ExperimentsClient client;
        private String name;
        private String description;
        private Dataset dataset;
        private Task<I, O> task;
        private final Map<String, Evaluator<?>> evaluators = new LinkedHashMap<>();
        private Map<String, Object> config;
        private Map<String, String> tags;

        Builder(ExperimentsClient client) {
            this.client = Objects.requireNonNull(client, "client");
        }

        public Builder<I, O> name(String name) {
            this.name = name;
            return this;
        }

        public Builder<I, O> description(String description) {
            this.description = description;
            return this;
        }

        public Builder<I, O> dataset(Dataset dataset) {
            this.dataset = dataset;
            return this;
        }

        public Builder<I, O> task(Task<I, O> task) {
            this.task = task;
            return this;
        }

        public Builder<I, O> evaluator(String label, Evaluator<?> evaluator) {
            Objects.requireNonNull(label, "evaluator label");
            Objects.requireNonNull(evaluator, "evaluator");
            this.evaluators.put(label, evaluator);
            return this;
        }

        public Builder<I, O> config(Map<String, Object> config) {
            this.config = config;
            return this;
        }

        public Builder<I, O> tags(Map<String, String> tags) {
            this.tags = tags;
            return this;
        }

        public Experiment<I, O> build() {
            Objects.requireNonNull(name, "name is required");
            Objects.requireNonNull(dataset, "dataset is required");
            Objects.requireNonNull(task, "task is required");
            return new Experiment<>(this);
        }
    }
}

File: experiments/node-sdk/src/Experiment.ts
import type { Dataset } from "./Dataset.js";
import type { Evaluator } from "./Evaluator.js";
import { ExperimentResult, Row } from "./ExperimentResult.js";
import type { ExperimentsClient } from "./ExperimentsClient.js";
import type { Task } from "./Task.js";
import * as http from "./internal/http.js";
import { appBase } from "./internal/http.js";
import { hexId } from "./internal/ids.js";
import { toMetric } from "./internal/metricBuilder.js";
import { toSpan } from "./internal/spanBuilder.js";

/**
 * Builder + `run()` orchestration.
 *
 * v0.1 runs sequentially (one row at a time), emits one root span per dataset
 * row, and posts all spans + metrics in a single events call.
 */
export class Experiment<I = unknown, O = unknown> {
  private readonly client: ExperimentsClient;
  private readonly _name: string;
  private readonly _description: string;
  private readonly dataset: Dataset;
  private readonly task: Task<I, O>;
  private readonly evaluators: Map<string, Evaluator>;
  private readonly config: Record<string, unknown>;
  private readonly tags: Record<string, string>;
  private _experimentId: string | null = null;

  private constructor(b: ExperimentBuilder<I, O>) {
    if (!b._name) throw new Error("Experiment name is required");
    if (!b._dataset) throw new Error("Experiment dataset is required");
    if (!b._task) throw new Error("Experiment task is required");
    this.client = b.client;
    this._name = b._name;
    this._description = b._description;
    this.dataset = b._dataset;
    this.task = b._task;
    this.evaluators = new Map(b._evaluators);
    this.config = { ...b._config };
    this.tags = { ...b._tags };
  }

  static builder<I = unknown, O = unknown>(client: ExperimentsClient): ExperimentBuilder<I, O> {
    return new ExperimentBuilder<I, O>(client);
  }

  /** @internal — bridge so ExperimentBuilder.build() can reach the private ctor. */
  static fromBuilder<I, O>(b: ExperimentBuilder<I, O>): Experiment<I, O> {
    return new Experiment<I, O>(b);
  }

  name(): string {
    return this._name;
  }

  experimentId(): string | null {
    return this._experimentId;
  }

  /** Dashboard URL for the experiment. Null until `run()` has created it. */
  url(): string | null {
    if (this._experimentId === null) return null;
    return `${appBase(this.client.site())}/llm/experiments/${this._experimentId}`;
  }

  /** Run the experiment sequentially and return the result. */
  async run(): Promise<ExperimentResult> {
    const projectId = await this.client.ensureProjectId();

    // Ensure the dataset exists remotely and all records are pushed.
    await this.dataset.ensureCreatedAndPushed(projectId);
    const datasetId = this.dataset.id();
    if (datasetId === null) {
      throw new Error(`Dataset '${this.dataset.name()}' has no id after push`);
    }

    // Create the experiment through the generated client. ensureUnique:true makes
    // the backend mint a fresh experiment under the project on every run.
    try {
      const resp = await this.client.api().createLLMObsExperiment({
        body: {
          data: {
            type: "experiments",
            attributes: {
              name: this._name,
              projectId,
              datasetId,
              description: this._description,
              ...(Object.keys(this.config).length > 0 ? { config: this.config } : {}),
              ensureUnique: true,
            },
          },
        },
      });
      this._experimentId = String(resp?.data?.id ?? "");
    } catch (err) {
      throw new Error(`Failed to create experiment '${this._name}': ${(err as Error).message}`);
    }
    const experimentId = this._experimentId as string;

    // Best-effort interrupt handling: if the process is signalled mid-run, mark
    // the experiment interrupted.
    let finished = false;
    const interruptHandler = () => {
      if (!finished) {
        void this.updateStatus("interrupted", "Process terminated before run completed");
      }
    };
    process.once("SIGINT", interruptHandler);
    process.once("SIGTERM", interruptHandler);

    try {
      const records = this.dataset.records();
      const recordIds = this.dataset.recordIds();
      const rows: Row[] = [];
      const spans: Record<string, unknown>[] = [];
      const metrics: Record<string, unknown>[] = [];

      for (let i = 0; i < records.length; i++) {
        const record = records[i];
        const datasetRecordId = i < recordIds.length ? recordIds[i] : "";
        const spanId = hexId();
        const traceId = hexId();
        const startMs = Date.now();
        const startNs = startMs * 1_000_000;
        const startHr = process.hrtime.bigint();

        let output: unknown = null;
        let errorType: string | null = null;
        let errorMessage: string | null = null;

        try {
          output = await this.task(record.input as I, this.config);
        } catch (err) {
          const e = err as Error;
          errorType = e.name || "Error";
          errorMessage = e.message ?? String(err);
        }

        const durationNs = Number(process.hrtime.bigint() - startHr);

        // Run evaluators (only when the task itself did not error).
        const evaluations: Record<string, unknown> = {};
        const evaluationErrors: Record<string, string> = {};
        const timestampMs = Date.now();

        for (const [label, evaluator] of this.evaluators) {
          if (errorType !== null) {
            // Task failed — record an evaluation error so the metric reflects it.
            const msg = "task error; evaluation skipped";
            evaluationErrors[label] = msg;
            metrics.push(
              toMetric(label, null, msg, spanId, timestampMs, experimentId, this.tags),
            );
            continue;
          }
          try {
            const value = await evaluator(record.input, output, record.expectedOutput);
            evaluations[label] = value;
            metrics.push(
              toMetric(label, value, null, spanId, timestampMs, experimentId, this.tags),
            );
          } catch (err) {
            const msg = (err as Error).message ?? String(err);
            evaluationErrors[label] = msg;
            metrics.push(
              toMetric(label, null, msg, spanId, timestampMs, experimentId, this.tags),
            );
          }
        }

        const row = new Row({
          index: i,
          spanId,
          traceId,
          startNs,
          durationNs,
          input: record.input,
          output,
          expectedOutput: record.expectedOutput,
          errorType,
          errorMessage,
          evaluations,
          evaluationErrors,
        });
        rows.push(row);
        spans.push(
          toSpan(
            row,
            record.metadata,
            experimentId,
            projectId,
            datasetId,
            datasetRecordId,
            this._name,
            this.tags,
          ),
        );
      }

      // Post all spans + metrics in a single events call.
      await this.postEvents(experimentId, spans, metrics);
      await this.updateStatus("completed", null);
      finished = true;

      return new ExperimentResult(experimentId, rows, this.url() as string);
    } catch (err) {
      await this.updateStatus("failed", (err as Error).message ?? String(err));
      finished = true;
      throw err;
    } finally {
      process.removeListener("SIGINT", interruptHandler);
      process.removeListener("SIGTERM", interruptHandler);
    }
  }

  private async postEvents(
    experimentId: string,
    spans: Record<string, unknown>[],
    metrics: Record<string, unknown>[],
  ): Promise<void> {
    // W2: the events POST uses type "experiments" (not "events").
    const body = {
      data: {
        type: "experiments",
        attributes: { spans, metrics },
      },
    };
    await http.post(
      this.client.credentials(),
      `/api/v2/llm-obs/v1/experiments/${experimentId}/events`,
      body,
    );
  }

  private async updateStatus(status: string, error: string | null): Promise<void> {
    // Hand-rolled: the generated updateLLMObsExperiment model exposes only
    // name/description, not `status`/`error`, so the lifecycle PATCH bypasses it.
    if (this._experimentId === null) return;
    const attributes: Record<string, unknown> = { status };
    if (error !== null) attributes.error = error;
    try {
      await http.patch(
        this.client.credentials(),
        `/api/v2/llm-obs/v1/experiments/${this._experimentId}`,
        { data: { type: "experiments", attributes } },
      );
    } catch {
      // Status update is best-effort; never let it mask the real result/error.
    }
  }
}

/** Builder for {@link Experiment}. */
export class ExperimentBuilder<I = unknown, O = unknown> {
  /** @internal */ readonly client: ExperimentsClient;
  /** @internal */ _name = "";
  /** @internal */ _description = "";
  /** @internal */ _dataset: Dataset | null = null;
  /** @internal */ _task: Task<I, O> | null = null;
  /** @internal */ _evaluators = new Map<string, Evaluator>();
  /** @internal */ _config: Record<string, unknown> = {};
  /** @internal */ _tags: Record<string, string> = {};

  constructor(client: ExperimentsClient) {
    this.client = client;
  }

  name(name: string): this {
    this._name = name;
    return this;
  }

  description(description: string): this {
    this._description = description;
    return this;
  }

  dataset(dataset: Dataset): this {
    this._dataset = dataset;
    return this;
  }

  task(task: Task<I, O>): this {
    this._task = task;
    return this;
  }

  /** Register an evaluator under a label. Multiple calls allowed. */
  evaluator(label: string, evaluator: Evaluator): this {
    this._evaluators.set(label, evaluator);
    return this;
  }

  config(config: Record<string, unknown>): this {
    this._config = { ...config };
    return this;
  }

  tags(tags: Record<string, string>): this {
    this._tags = { ...tags };
    return this;
  }

  build(): Experiment<I, O> {
    return Experiment.fromBuilder<I, O>(this);
  }
}
